src/collection/tx_data/collector.py

```diff
@@ -0,0 +1,282 @@
+
+    async def _async_fetch_block(self, block_number: int) -> Tuple[UiConfirmedBlock | None, int]:
+        """
+        Use solana client to fetch block with provided number.
+        """
+        await self._async_rate_limit_calls()
+        # Fetch block data.
+        try:
+            block = await self.async_solana_client.get_block(
+                slot=block_number,
+                encoding='jsonParsed',
+                max_supported_transaction_version=0,
+            )
+        except SolanaRpcException as e:
+            LOG.error(f"SolanaRpcException while fetching {block_number}: {e}")
+            await asyncio.sleep(0.5)
+            return await self._async_fetch_block(block_number)
+        except RPCException as e:
+            LOG.error(f"RpcException while fetching {block_number}: {e}")
+            return None, block_number
+        except SerdeJSONError as e:
+            tb_str = traceback.format_exc()
+            # Log the error message along with the traceback
+            LOG.error(f"An error occurred: {e}\nTraceback:\n{tb_str}")
+            await asyncio.sleep(0.3)
+            return await self._async_fetch_block(block_number)
+
+        return block.value, block_number
+
+    def _select_relevant_tx_from_block(self, block: UiConfirmedBlock, block_number: int) -> None:
+        """
+        Select only relevant transactions based on public keys involved.
+        """
+        if block.transactions:
+            relevant_transactions = [tx for tx in block.transactions if self._is_transaction_relevant(tx)]
+            # Store raw transaction data with block metadata as `SolanaTransaction` objects.
+
+            relevant_solana_transaction = [
+                SolanaTransaction(
+                    block_number=block_number,
+                    block_time=block.block_time if block.block_time else -1,
+                    tx_body=tx
+                ) for tx in relevant_transactions  # type: ignore
+            ]
+            self.rel_transactions.extend(relevant_solana_transaction)
+
+    @log_performance_time(LOG)
+    def _get_assignment(self) -> None:
+        """
+        Obtain assignment for data collection.
+        """
+        # Update for relevant PPKs
+        self._get_protocol_public_keys()
+        # receive block numbers ready for collection
+        self._get_assigned_blocks()
+        LOG.info(f"Blocks for collection: {self.assignment}")
+
+    @abstractmethod
+    def _get_assigned_blocks(self) -> None:
+        """
+        Obtains block numbers to fetch.
+        """
+        raise NotImplementedError("Implement me!")
+
+    def _get_protocol_public_keys(self) -> None:
+        """
+        Get list of public keys for relevant protocol from env variables.
+        :return:
+        """
+        if not self.protocol_public_keys:
+            self.protocol_public_keys = list()
+        # get list of public keys from env variables.
+        keys = os.getenv("PROTOCOL_PUBLIC_KEYS", "").split(',')
```

Comment:
What's the advantage of keeping this as an env variable instead of having it in settings?

Reply:
I have the same question for `RATE_LIMIT`.

Reply:
so we have an easy way to setup containers with different rate limits. when historical collection is done, rate limit can be easily reallocate to current collector by restarting container with different `-e RATE_LIMIT`

same for ppks, for instance if we realize that we data for wrong ppk or ppk is missing.

Reply:
done. 9fad6d5669f05c62fc65c17680c1ae08a8327620
i kept using env to obtain PPKs as an option.